Operator: add support for downscaling

nicolaferraro opened the pull request:
## Cover letter

This adds support for downscaling clusters in Kubernetes. Downscaling is done in coordination with the cluster, using the `/decommission` endpoint. The StatefulSet is downscaled only if the cluster allows it (e.g. having a topic with 3 replicas prevents the cluster to scale to less than 3 nodes).

When the cluster prevents the statefulset to scale, replicas can still be reverted to previous state and the operator will automatically trigger a `/recommission` of the broker.

This also removes the need to use an empty seed_server when node 0 starts, because fresh clusters now start with a single replica, then are upscaled.

~The whole PR has been created with `nodeID` possibly diverging from StatefulSet `ordinal` (a PR to do that is expected after this).~

It also deals with many edge cases (see controller tests).

## Release notes

### Features

* operator: clusters can now be downscaled (alpha feature enabled using the controller startup flag `--allow-downscaling`)
* operator: multi-replica clusters are initialized with one replica then upscaled to the desired number, without need to create empty seed-server

nicolaferraro commented:
> Code ready for review.
> 
> It needs #4964 to work properly, because when scaling up from one node, node 0 is restarted and sometimes the cluster fails to get leadership.

The problem with node 0 not being able to form the initial cluster after restart was caused by maintenance mode hooks enabled on it: i.e. node 0 was losing leadership and no longer be leader for anything after restart. I've disabled maintenance mode when the cluster is starting up. That forces a restart during multi-node cluster first creation (it will be fixed by #4907)

pvsune commented on src/go/k8s/pkg/resources/statefulset.go:
```diff
@@ -269,6 +285,10 @@ func (r *StatefulSetResource) obj(
 	tlsVolumes, tlsVolumeMounts := r.volumeProvider.Volumes()
+
+	// We set statefulset replicas via status.replicas in order to control it from the handleScaling function
+	replicas := r.pandaCluster.Status.GetReplicas()
```
Nit: I think normally the operator will update the desired state of StatefulSet then this will reconcile back to the owner which is the Cluster.

nicolaferraro replied:
Yeah, it's a bit unconventional. The problem is that I needed a field in the status to tell all individual reconcilers how to behave. E.g. the current replicas changes the way we populate both statefulset and configmap, but also how we contact the cluster via admin API etc.

So I needed a field to send an input to the other controllers, containing the decision on how many instances to run. I took a shortcut and used `status.replicas` instead of something like `status.desiredReplicas`, since they would have matched at next reconcile loop..

alenkacz replied:
yeah maybe we should rename to avoid confusion, I agree the convention is to use replicas a bit differently and I was also confused

pvsune commented on src/go/k8s/pkg/resources/statefulset.go:
```diff
@@ -269,6 +285,10 @@ func (r *StatefulSetResource) obj(
 	tlsVolumes, tlsVolumeMounts := r.volumeProvider.Volumes()
+
+	// We set statefulset replicas via status.replicas in order to control it from the handleScaling function
+	replicas := r.pandaCluster.Status.GetReplicas()
```
Nit: I think normally the operator will update the desired state of StatefulSet then this will reconcile back to the owner which is the Cluster.

nicolaferraro commented on src/go/k8s/pkg/resources/statefulset_scale.go:
```diff
@@ -0,0 +1,420 @@
+
+// handleScaling is called to detect cases of replicas change and apply them to the cluster
+// nolint:nestif // for clarity
+func (r *StatefulSetResource) handleScaling(ctx context.Context) error {
```
I get your point. StatefulSets manage certain aspects of the lifecycle of pods, but not others. This is an old battle that I fighted in the past, without success: https://github.com/kubernetes/kubernetes/issues/59343

In general, I think that we should leave basic pod management to the statefulset controller and implement the remaining things, like decommission on top. The `handleScaling` function does not control pods directly, it controls the `spec.replicas` of the StatefulSet, then relies on the statefulset controller to do the remaining job.

alenkacz commented on src/go/k8s/pkg/resources/statefulset_scale.go:
```diff
@@ -0,0 +1,392 @@
+}
+
+// isRunningReplicas checks if the statefulset is configured to run the given amount of replicas and that also pods match the expectations
+func (r *StatefulSetResource) isRunningReplicas(
```
`verifyRunningCount` ? 🤔

nicolaferraro commented on src/go/k8s/pkg/resources/statefulset_scale.go:
```diff
@@ -0,0 +1,392 @@
+
+// guessNodeID tries to get the node_id from the broker or falls back to guessing it by querying other nodes,
+// returning nil in case the node might not have one
+func (r *StatefulSetResource) guessNodeID(
```
Looking at the conversations now, I guess we'll wait for core to implement reusable node IDs. So I'll remove this part and use the ordinal only.

alenkacz replied:
oh I see, it was preparation for that. Got it 👍

nicolaferraro commented on src/go/k8s/controllers/redpanda/cluster_controller.go:
```diff
@@ -353,7 +353,7 @@ func statusShouldBeUpdated(
 			!reflect.DeepEqual(nodeList.SchemaRegistry, status.Nodes.SchemaRegistry) ||
 			!reflect.DeepEqual(nodeList.ExternalBootstrap, status.Nodes.ExternalBootstrap)) ||
-		status.Replicas != sts.LastObservedState.Status.ReadyReplicas ||
+		status.ReadyReplicas != sts.LastObservedState.Status.ReadyReplicas ||
```
Ok, I'll treat `replicas` as a mirrored field, and add some other, like `desiredReplicas` or `targetReplicas`, to indicate the decision.

nicolaferraro commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_types.go:
```diff
@@ -344,6 +347,14 @@ const (
+	// Ordinal is the progressive number corresponding to the node being decommissioned
+	Ordinal int32 `json:"ordinal,omitempty"`
+	// NodeID is the Cluster internal identifier of the node being decommissioned
+	NodeID *int `json:"nodeID,omitempty"`
```
Technically a nil indicates a direct decommission. But you made a good point about unifying the two phases that I think I can remove it altogether and replace the struct with a single field (`decommissioningNode`)

nicolaferraro commented on src/go/k8s/pkg/resources/statefulset.go:
```diff
@@ -167,8 +167,15 @@ func (r *StatefulSetResource) Ensure(ctx context.Context) error {
+	}
+
+	r.logger.Info("Running scale handler", "resource name", r.Key().Name)
+	return r.handleScaling(ctx)
```
Mmh.. Here the update takes precedence on the scaling. The statefulset_update handler function will throw errors until the update is done, so scaling is done only after update.

An update in the middle of scaling will interrupt it.. until done

alenkacz replied:
do we have a test for this? Can we maybe change one of the updat test to also scale to see both can be done at the same time?

nicolaferraro replied:
Yeah, I'm thinking to a specific test about this...

nicolaferraro replied:
Not easy to test.. I'd need to mock statefulset_update as well, because it's doing direct API calls...

nicolaferraro commented on src/go/k8s/pkg/resources/statefulset_scale.go:
```diff
@@ -0,0 +1,392 @@
+	if r.pandaCluster.Status.Decommission != nil {
+		if r.pandaCluster.Status.Decommission.NodeID == nil {
+			// Special type of decommission for nodes that probably never started
+			return r.handleDirectDownscale(ctx)
```
At the end of the direct scale down there's a check that the node is not registered. If the node is found, then proper decommission happens

nicolaferraro commented on src/go/k8s/pkg/resources/statefulset_scale.go:
```diff
@@ -0,0 +1,392 @@
+	targetReplicas := r.pandaCluster.Status.Decommission.Ordinal
+	r.logger.Info("Handling cluster in decommissioning phase", "target replicas", targetReplicas)
+
+	fullAdminAPI, err := r.getAdminAPIClient(ctx)
```
There are cases where I use an admin API client that can reach every node of the cluster and other cases where I target a specific node to get the node ID. I guess it will become just adminAPI once I remove the distinction between nodeId and ordinal

nicolaferraro commented on src/go/k8s/pkg/resources/statefulset_scale.go:
```diff
@@ -0,0 +1,392 @@
+	}
+
+	// Broker is now missing from cluster API
+	r.logger.Info("Node has been successfully decommissioned from cluster", "node_id", *r.pandaCluster.Status.Decommission.NodeID)
```
Process still running (also verified with manual calls). I shut it down when scaling the statefulset down

nicolaferraro commented on src/go/k8s/pkg/resources/statefulset_scale.go:
```diff
@@ -0,0 +1,392 @@
+		return r.Status().Update(ctx, r.pandaCluster)
+	}
+
+	scaledDown, err := r.isRunningReplicas(ctx, targetReplicas)
```
Yes, it makes perfect sense. I can skip the decommission part if the broker is not there. I can add an extra check that is ok for both cases.. unifying them

:+1:

nicolaferraro commented:
This is ready for a second pass. Other than the suggestions, I've also found other issues that I've fixed in this new version:
- When upgrading the operator and there's an existing cluster running, it's likely that `status.desiredReplicas` of the CR is not initialized (unlike the old `status.replicas`), so I took care of initializing it properly for existing clusters. Only fresh clusters do the initialization starting from 1 node (then scaling up), while existing clusters are initialized at the current number of replicas
- When a node is decommissioned and shut down, its maintenance mode hooks can put the cluster in an inconsistent state (node deleted from the cluster, but at the same time holding the lock on maintenance mode, preventing other nodes to enter maintenance mode and shutdown properly). See: #4999. I've added a workaround to force releasing the maintenance mode lock
- I've added a hack to still remove the seeds servers when they contain a single entry. Normally clusters are able to form when the seeds server list contain a single entry pointing to themselves, except when the Kafka API is secured with mutual TLS (this is why the helm test was failing). So I empty the seeds server list when it contains a single entry: it should be equivalent and still prevent the issues that we're facing with the logic based on ordinal 0..

cc: @alenkacz , @pvsune, @RafalKorepta

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_types.go:
```diff
@@ -847,6 +866,30 @@ func (s *ClusterStatus) SetRestarting(restarting bool) {
+// ComputeInitialDesiredReplicas calculates the initial value for status.desiredReplicas.
+//
+// It needs to consider the following cases:
+// - Fresh cluster: we start from 1 replicas, then upscale if needed
```
Checking HPA docs, I wonder whether this field should not be named currentreplicas rather? I feel like desired is the target and final one 🤔 https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/

but feel free to correct me, I just want to get the API right so it's not confusing since it's hard to change

nicolaferraro replied:
Yeah, I took the name from HPA, but I agree it's confusing, especially on the part where the user "desires" multiple replicas, but the controller keeps them to a lower value. `currentReplicas` may be better..

nicolaferraro replied:
The reason for 1 replica is to prevent the problem with seeds_servers, to have them empty only at cluster creation and never again

alenkacz commented on src/go/k8s/cmd/configurator/main.go:
```diff
@@ -131,9 +131,11 @@ func main() {
+	// In case of a single seed server, the list should contain the current node itself.
+	// Normally the cluster is able to recognize it's talking to itself, except when the cluster is
+	// configured to use mutual TLS.
+	// So, we clear the list of seeds to help Redpanda.
```
do you know why is that? the nodes does not talk to itself by mtls. Is this a bug in redpanda they are looking to fix at one point?

nicolaferraro replied:
https://github.com/redpanda-data/redpanda/issues/333#issuecomment-1153927714

alenkacz commented on src/go/k8s/controllers/redpanda/suite_test.go:
```diff
@@ -365,12 +382,114 @@ func (m *mockAdminAPI) SetUnavailable(unavailable bool) {
+	return admin.NodeConfig{}, nil
+}
+
+// nolint:goerr113 // test code
```
we need to disable this linter completely 🙄

alenkacz commented on src/go/k8s/pkg/resources/statefulset.go:
```diff
@@ -166,8 +175,21 @@ func (r *StatefulSetResource) Ensure(ctx context.Context) error {
 	}
 	r.LastObservedState = &sts
+
+	// Hack for: https://github.com/redpanda-data/redpanda/issues/4999
```
uff :)

alenkacz reviewed (commented):
I think the API is the only blocker for me, otherwise looks REALLY good, thank you 👏

RafalKorepta commented on src/go/k8s/pkg/resources/statefulset_scale.go:
```diff
@@ -261,6 +263,56 @@ func (r *StatefulSetResource) isClusterFormed(
+	ordinal := *r.pandaCluster.Status.DecommissioningNode
+	targetReplicas := ordinal
+
+	scaledDown, err := r.verifyRunningCount(ctx, targetReplicas)
```
NIT: The verify running count doesn't check if targetReplicas aka decommissioning node id is decreasing by looking at statefulset definition

```suggestion
	differentCount, err := r.verifyRunningCount(ctx, targetReplicas)
```

RafalKorepta commented on src/go/k8s/pkg/resources/statefulset_scale.go:
```diff
@@ -261,6 +263,56 @@ func (r *StatefulSetResource) isClusterFormed(
+// preventing other pods clean shutdown.
+//
+// See: https://github.com/redpanda-data/redpanda/issues/4999
+func (r *StatefulSetResource) disableMaintenanceModeOnDecommissionedNodes(
```
Any unit test/ginko test?

RafalKorepta reviewed (commented):
Thank you for this amazing set of features! I left few comments.

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_types.go:
```diff
@@ -841,6 +841,16 @@ func (r *Cluster) IsUsingMaintenanceModeHooks() bool {
 
+// ClusterSpec
+
+// GetReplicas returns the replicas field is present, or 1
```
I guess so but I think it's a sane default... it's either 1 or 0 🤷‍♀️

RafalKorepta replied:
Disallow empty replicas

alenkacz reviewed (approved):
LGTM

RafalKorepta commented on src/go/k8s/pkg/resources/statefulset_scale.go:
```diff
@@ -0,0 +1,255 @@
+	}
+
+	if broker == nil || broker.MembershipStatus != admin.MembershipStatusActive {
+		err = adminAPI.RecommissionBroker(ctx, int(*r.pandaCluster.Status.DecommissioningNode))
```
It became almost instantaneously active.

RafalKorepta commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_types_test.go:
```diff
@@ -216,3 +217,19 @@ func TestConditions(t *testing.T) {
+	assert.Equal(t, int32(1), cluster.GetCurrentReplicas())
+	cluster.Status.Replicas = 2
+	assert.Equal(t, int32(3), cluster.GetCurrentReplicas())
+	cluster.Status.Replicas = 0
+	cluster.Status.ReadyReplicas = 2
```
It was done to make sure function is working correctly. In the real world Replicas will not be lower than ReadyReplicas

nicolaferraro commented:
Done latest changes we agreed:
- Removed possibility to set nil replicas (was not my intention). Now it's enforced in the webhook, previously the cluster just failed on nil replicas
- Made decommission wait interval configurable (plus jitter, no backoff as I needed to track it per CR)

RafalKorepta commented on src/go/k8s/pkg/resources/statefulset_scale.go:
```diff
@@ -0,0 +1,392 @@
+	}
+
+	if r.pandaCluster.Status.DecommissioningNode == nil || r.pandaCluster.Status.CurrentReplicas > *r.pandaCluster.Status.DecommissioningNode {
+		// Only if actually in a decommissioning phase
```
NIT: This comment miss-leads as it is the opposite.  When entered it means do nothing as decommissioning is not executed or it is scaling up. 

This comment would be more accurate 2 lines down.

RafalKorepta replied:
```suggestion
		// Only if not in decommissioning phase
```

RafalKorepta commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook.go:
```diff
@@ -43,6 +43,11 @@ const (
 	defaultSchemaRegistryPort = 8081
 )
 
+// AllowDownscalingInWebhook controls the downscaling alpha feature in the Cluster custom resource.
+// Downscaling is not stable since nodeIDs are currently not reusable, so adding to a cluster a node
+// that has previously been decommissioned can cause issues.
```
It would be good to describe in commit message or in this comment what consequences might happen if someone will downscale the cluster while Kafka clients are still connect. 

cc @jcsp @mmaslankaprv

RafalKorepta reviewed (approved):
I left none blocking comments as they are only stylish NITs or documentation.

Is there planed work on documentation?

cc @Feediver1
